cmake/ProhibitInSourceBuild.cmake

```diff
@@ -0,0 +1,23 @@
+
+# Distributed under the University of Illinois/NCSA Open Source License.
+# See LICENSE.txt for details.
+
+function(prohibit_in_source_build)
+  if ("${CMAKE_BINARY_DIR}" STREQUAL "${CMAKE_SOURCE_DIR}")
+    message(FATAL_ERROR "\n"
+      "You attempted to build ${PROJECT_NAME} in the directory:\n"
+      "  ${CMAKE_SOURCE_DIR}\n"
```

Comment:
This should probably be `${CMAKE_BUILD_DIR}` if we also end up blocking `src` builds then having this be `${CMAKE_SOURCE_DIR}` would give a confusing error